### DISCUSSION TITLE
guarantee FOSS licensing of SixtyFPS

### DISCUSSION THREAD

**Discussion by Be-ing** [slint-ui/slint#244] [General] [answered]
I'm really interested and excited for what SixtyFPS will become, but I'm also very wary of using it, promoting it, or contributing to it because of the [CLA](https://cla-assistant.io/sixtyfpsui/sixtyfps), especially in the context of [recent events](https://github.com/audacity/audacity/discussions/932).

> To the maximum extent permitted by the relevant law, You grant to Us a perpetual, worldwide, non-exclusive, transferable, royalty-free, irrevocable license under the Copyright covering the Contribution, with the right to sublicense such rights through multiple tiers of sublicensees, to reproduce, modify, display, perform and distribute the Contribution as part of the Material; provided that this license is conditioned upon compliance with Section 2.3.

This is not a fair exchange. There is no assurance that SixtyFPS won't become completely proprietary in the future and new versions of SixtyFPS will become unavailable to application developers using it with the GPL license. I don't like giving SixtyFPS permission to use code I'd contribute as proprietary software, but it would be tolerable if there was a legal commitment that SixtyFPS will always be available as FOSS like the [KDE Free Qt Foundation's agreement with the Qt Company](https://kde.org/community/whatiskde/kdefreeqtfoundation/). Would you consider adding a clause to the CLA or some other legal commitment that SixtyFPS and future versions of it will always be available as FOSS? I hope as old KDE and Qt developers you understand how important this is for building a community around SixtyFPS.

**Comment by tronical** [comment]
Hi @Be-ing,

our roots as software developers are in the open source community and we are interested in finding a way to express our commitment to there always being an open source version available. We are open to expressing this in the CLA and more publicly, in the absence of an entity like the KDE Free Qt Foundation. We'll draft something and post it here for review.

**Reply by Be-ing** [reply]
I suggest adding a clause stating something along the lines of "In addition to the above license, Contribution and derivative works of Contribution will be published by SixtyFPS under the terms of the GNU General Public License, version 3." Alternatively, if you want it to be more flexible, you may consider saying it will be published with a license approved by the OSI which is compatible with GPLv3 (to be clear that GPLv3 applications can continue to link to SixtyFPS). I'll leave the exact wording to you and your lawyer.

By the way, is SixtyFPS licensed GPLv3 only or GPLv3 or later? I suppose it doesn't matter much with the CLA because you could relicense with later versions of the GPL already with the existing wording of the CLA. Personally, my feelings about whether the "or later" term is a good idea are mixed considering recent events at the FSF.

**Reply by Be-ing** [reply]
Another example to consider is [Signal's CLA](https://signal.org/cla/):

> Grant of Copyright License. Subject to the terms and conditions of this Agreement, You hereby grant to Open Whisper Systems and to recipients of software distributed by Signal Messenger a perpetual, worldwide, non-exclusive, no-charge, royalty-free, irrevocable copyright license to reproduce, prepare derivative works of, publicly display, publicly perform, and distribute Your Contributions and such derivative works, as well as the right to sublicense and have sublicensed all of the foregoing rights, through multiple tiers of sublicensees, provided that in all cases, Signal Messenger will make Your Contributions available under an OSI-approved open source license.

**Reply by tronical** [reply]
Hi everyone,

We've been pondering about this and we've also gotten some advice from a lawyer (who also happens to be very familiar with the KDE Free Qt Foundation setup). We identify three aspects to this:

 1. Our intentions of creating, growing and participating in an open source community, while establishing a sustainable business - built on the idea that there are users out there who choose not to participate in the free software ecosystem and are willing to pay a fee in exchange that helps fund the development.

 2. A technical mechanism that ensures that accepted contributions are immediately available to the community.

 3. A legal framework that more generically ensures the future availability of contributions under an open source license.

If we misbehave with regards to our intentions, if we fail to participate in the community and fail to respect the effort contributors put in, then - in this early phase - it is more than plausible that this may have a negative impact on the business, too. Conversely, a thriving community can significantly contribute to the sustainability of a business. It's one of the special aspects of a software developer product: it's code contributors are also its users.

Now, technical and legal mechanisms can be put in place to support this and further strengthen trust between contributors and the people with the additional rights to hand out commercial licenses. But no matter how strong those mechanisms, once the intentions and behavior change, the mechanisms can't restore the trust that is lost. This may prevent the community from thriving and growing jointly with the license holders. In the best case, momentum suffices to revive the community around a fork. In the worst case, the project slowly fades out and becomes irrelevant - from the perspective of the world of open source software as well as the business, likely.


There is a technical mechanism in place today that is controlled by an external entity (GitHub):

 1. Every contribution through a merge request is publicly visible. Its origin is in a repository fork.
 2. Once accepted, it is instantly distributed under the terms of the GPLv3 in the sixtyfps repository.

This mechanism is working well - as it is shared with thousands of other projects on GitHub - and the transparency of it is refreshing.

We do not have an additional legal mechanism in place today and we are finding it difficult to formulate this in the CLA in an unambiguous and clear way. For example a clause like in the Signal CLA seems like it would make it impossible to deliver a patch that is specific to a proprietary customer's environment.

Otherwise we are short of an independent third-party that we could make an agreement with. This is however something we are interested in in the long term. Meanwhile, we would like to proceed with writing down our intentions as a pledge, and we are happy with the infrastructure that GitHub provides to everyone.

This is a draft of the kind of pledge that we'd like to put into a more prominent place then:

> We believe that open source software development and communities are the foundation for a healthy ecosystem of high-quality software, where everyone can learn, improve and give
> back. We commit to upholding this foundation and pledge by promising to continue to develop SixtyFPS in the open under an open source license.

What do you think?

**Reply by Be-ing** [reply]
Thank you for this detailed response. It demonstrates that you do care about doing right by the community. I agree that actions speak louder than words and trust is more important than legal documents. At the same time, having legal commitments in place that communicate your intentions is helpful for building that trust. The draft pledge you posted is better than nothing but not legally enforceable. While I might trust you now, what would happen to that legally unenforceable commitment if SixtyFPS gets bought by another company?

> We do not have an additional legal mechanism in place today and we are finding it difficult to formulate this in the CLA in an unambiguous and clear way. For example a clause like in the Signal CLA seems like it would make it impossible to deliver a patch that is specific to a proprietary customer's environment.

I am not a lawyer so maybe I am missing something here, but I don't understand the contradiction. If you make a proprietary patch for a specific customer's request and only distribute it to that customer, the CLA doesn't apply to that patch, does it? If the patch is generally useful beyond a customer's niche scenario, then I think it should go in the public repository. That's more practical to maintain anyway from a technical perspective.

**Reply by IndianTurkishCLA** [reply]
I just want to say that signing CLA makes me think, why? Why would i sign anything to contribute in open source and take something from that open source? Should i? My brain says no, but i also understand that law and another shitshow can pop-up in some moments, but who care though? Nobody. Also, anyone who have at least not room temperature digits can understand that it basically means you taking some responsibility, can't they?
Would you sign a letter to pay 5$ for buying some food?
What about holding your wife's/girlfriend's clutch while she's doing shop?
Would you sign something to help in a common situation, for example, open a door for someone? 

But obviously you would sign something if you are some sort of medic or taking participate in some kind of life-saving situation, but only if you are not experienced enough (just to give guarantee that your involvement will not affect on that person's life, after success the sign should be destroyed, otherwise it puts you in weak position), but if you are good you need to think if you can control everything and everyone's actions to not ruin anything while you are doing your work, you don't need to sign anything if you are intelligent (and i'm not talking about IQ).

If you are weak or you have some mental issues (yes, i mean if you are experienced tweeter, github users) you would sign something.

It should be optional, in cases when the person wants - let him do that (trust me, we can even predict which ethnicities/races/experience level users will sign), if he doesn't want to - let them do whatever they want to.
Also let them choose if they want to stay in public contributors/thanks lists or no, it is also annoying when you contribute somewhere and those retards put your nickname in some sort of credits lists.
When i see some bug i just want to say hey, look, bug is here, i'm not going to sign 1000000 checkboxes and fill up my personal info to contribute, either i wouldn't put my info to get reward for reports or fixes.



![image](https://user-images.githubusercontent.com/87715984/126364057-7106e9da-d8f9-4e6e-93d8-7c68bc29a94e.png)
![image](https://user-images.githubusercontent.com/87715984/126364185-5916a02f-67a4-4141-8d1b-5e53b6971b15.png)
(the only sign people can get from me, dotted lines - never signed)

**Reply by IndianTurkishCLA** [reply]
Also i hope you all understand what it is supposed to be? Master -> main.
Linus Torvalds vs One of the linux core contributors (you could easily find out what i'm talking about), who was trying to contribute shitty code to core and constantly got declined by linus, guess what happened next? Yes, diversity. "The millenials got to him" (google).
Etc.
CLA yet another part of instrumentaries.

P.s i'm not even trolling.

**Reply by IndianTurkishCLA** [reply]
> We do not have an additional legal mechanism in place today and we are finding it difficult to formulate this in the CLA in an unambiguous and clear way. For example a clause like in the Signal CLA seems like it would make it impossible to deliver a patch that is specific to a proprietary customer's environment.

A little bit extended version of it.
Just write up some summaries about if you are using our software you are agreed with chance of getting a patch from random person that haven't sign CLA or doesn't want to sign CLA.


Let's work for quality and normal stable community, not for specific abused people. Otherwise we will continue with dumb shitty solutions with bugs and shitcode.


To understand everything above you need to undestand english, again, understand, if you can speak or write in english it doesn't mean that you can understand it. Process text, try to get meaning of words.

**Reply by er-vin** [reply]
> Otherwise we are short of an independent third-party that we could make an agreement with.

Did you try to bring it up to the KDE e.V. yet?

Has the advantage of the experience toward managing such a relationship. Also it'd raise awareness of this project towards the KDE community which can't be a bad thing.

**Reply by Be-ing** [reply]
Disregarding the incoherent rambling above, have you considered how you could make your commitment robust in the event that the SixtyFPS company is sold?

**Reply by tronical** [reply]
Hi Be-ing,

In full transparency, we are preparing a few changes to the CLA that includes an open source pledge. Below is the paragraph that we would like to add:

> ### 2.6 Our Obligation.
>￼
>￼We believe that open source software development and communities are the foundation for a healthy ecosystem of high-quality software, where everyone can learn, improve and give back. We commit to upholding this foundation and pledge by promising to continue to develop the Material in the open under an open source license. By making the Material publicly available via GitHub, we ensure that it – the Material – shall be available and accessible to the open source communities.

We will need to make another change since we just created the SixtyFPS GmbH company, and plan to do the two changes at the same time.


@er-vin: You have a point in that this would bring advantages. We haven't really contacted the KDE e.v. yet but we're thinking about it.

**Reply by Be-ing** [reply]
That looks good, although I think it would be good to define an "open source license" as one approved by the Open Source Initiative to ensure that the license can't be some [not-really-open-source license](https://sfconservancy.org/blog/2018/oct/16/mongodb-copyleft-drafting/) like MongoDB. Even better would be specifying that the open source license must allow combining SixtyFPS with GPL licensed software to ensure that GPL applications using SixtyFPS will be able to continue using it even if the open source license used for SixtyFPS changes.

I would remove the specific mention of GitHub because it could become awkward if you decide to move to different infrastructure in the future. Perhaps just remove the "via GitHub" from that sentence; I think "making the Material publicly available" would be sufficient.

**Reply by tronical** [reply, marked as answer]
Thanks for the feedback, both of your suggestions make sense. Updated proposal:

> ### 2.6 Our Obligation.
> 
> We believe that open source software development and communities are the foundation for a healthy ecosystem of high-quality software, where everyone can learn, improve and give back. We commit to upholding this foundation and pledge by promising to continue to develop the Material in the open under an open source license, as approved by the Open Source Initiative (https://opensource.org). By making the Material publicly available, we ensure that it – the Material – shall be publicly available in our  public source code repository and accessible to the open source communities.

**Reply by Be-ing** [reply]
That looks great, thanks. I'll definitely be keeping an eye on how this project develops and if I start any new GUI applications, this will be the first library I'll try.

**Reply by er-vin** [reply]
> @er-vin: You have a point in that this would bring advantages. We haven't really contacted the KDE e.v. yet but we're thinking about it.

@tronical that would be nice if you did. I fail to see how you can secure over time any guarantees without a trusted third party and a setup like the KDE Free-Qt Foundation otherwise. This is especially important for project/company take overs.

**Reply by Be-ing** [reply]
@tronical I don't see that [text](https://github.com/sixtyfpsui/sixtyfps/discussions/244#discussioncomment-1233772) in the [current CLA](https://cla-assistant.io/sixtyfpsui/sixtyfps).

**Reply by tronical** [reply]
We wanted to combine the CLA update with the registration of the company, but then this took longer than we expected. Then we forgot to update it :(. Apologies! It's updated now.

**Reply by Be-ing** [reply]
Thanks for taking care of that.
